v2/util/gen-templates.js

```diff
@@ -1,54 +1,81 @@
+  let postInfos = [];
+  const base = lang === 'en' ? enBase : cnBase;
   const source = `${base}atom.xml`;
   let rss = await http.get(source);
   const $ = cheerio.load(rss.data, { decodeEntities: false });
-  const html = fs.readFileSync('./util/posts-slide.html', 'utf-8');
-  const $Template = cheerio.load(html, { decodeEntities: false });
   const length = 10;
 
-  $Template('.swiper-wrapper').append('  ');
+  if (!picsExists) {
+    picsExists = await readdir('./images/header-images');
+  }
+
   $('entry').slice(0, length).each((index, entry) => {
     const title = $(entry).children('title').text();
     const array = $(entry).children('id').text().split('/');
     const id = array[array.length - 2];
     const pic = $(entry).children('pic').text();
-    const picName = pic.split('.')[0];
-    const webp = `${picName}.webp`;
-
-    const text = $Template(`<p class="blog-title">${title}</p>`);
-    const textContainer = $Template('<div class="article-text"></div>');
-    textContainer.append(text);
 
-    const source = $Template(`<source srcset="${base}${webp}" type="image/webp" />`);
-    const img = $Template(`<img class="article-img" src="${base}${pic}" />`);
-    const picture = $Template(`<picture></picture>`);
-    picture.append(source).append(img);
+    // images will be downloaded and compressed only when they do not exist yet
+    if (!picsExists.includes(pic.split('/header-images/')[1])) {
+      pics.add(pic);
+    }
 
-    const imgContainer = $Template(`<div class="blog-img" data-pic="${pic}"></div>`);
-    imgContainer.append(picture);
+    postInfos.push({href: `${base}${id}`, pic, title})
+  });
 
-    const link = $Template(`<a class="article-item" href="${base}${id}/?src=org" target="_blank">`);
-    link.append(imgContainer);
-    link.append(textContainer);
+  const compileFunction = pug.compileFile('./util/posts-slide.pug', {pretty: true});
+  writeFile(`./templates/posts-slide-${lang}.tt2`, compileFunction({postInfos}));
+};
 
-    const slide = $Template('<div class="swiper-slide"></div>');
-    slide.append(link);
+Promise.all([
+  genSlideTemplate('en'),
+  genSlideTemplate('cn'),
+]).then(() => {
+  for(pic of pics) {
+    optimizeImg(pic);
+  }
+})
 
-    $Template('.swiper-wrapper').append(slide);
-    $Template('.swiper-wrapper').append(index === length - 1 ? '\n      ' : '\n        ');
+async function optimizeImg(pic) {
+  const response = await axios({
+    url: `${cnBase}${pic}`,
+    method: "GET",
+    responseEncoding: "binary",
   });
 
-  fs.writeFileSync(`./templates/posts-slide-${lang}.tt2`, $Template.html() + '\n');
+  await writeFile(`./images${pic}`, response.data, 'binary');
+  compressImg(`./images${pic}`);
 };
 
-genSlideTemplate('en');
-genSlideTemplate('cn');
+async function compressImg(filePath) {
+  return sharp(filePath)
+    .resize({
+      height,
+      fit: 'contain',
+      withoutEnlargement: true,
+    })
+    .toBuffer()
+    .then(data => {
```

Comment:
Again, just use `await` in Async functions.

Reply:
changed.